### PR TITLE
websocket over http2

### PR DESCRIPTION
**Pull request by arenevier**
@ricea @jgraham

### CODE REVIEW

**Comment by arenevier** [comment]
This is the rfc for that feature https://github.com/web-platform-tests/rfcs/pull/67

File: tools/wptserve/wptserve/server.py
Comment (Hexcles) [inline tools/wptserve/wptserve/server.py:774]: The default for `ws_doc_root` should be `doc_root/websockets/handlers`

File: tools/wptserve/wptserve/server.py
Comment (Hexcles) [inline tools/wptserve/wptserve/server.py:467]: Nit: drop an empty line here.
```suggestion
```

File: tools/wptserve/wptserve/server.py
Comment (Hexcles) [inline tools/wptserve/wptserve/server.py:544]: same here

File: tools/wptserve/wptserve/ws_h2_handshake.py
Comment (Hexcles) [inline tools/wptserve/wptserve/ws_h2_handshake.py:18]: At least the exceptions are exposed in `mod_pywebsocket.handshake` via its `__init__.py`. We could do the same for other methods, too.

File: tools/wptserve/wptserve/server.py
Comment (Hexcles) [inline tools/wptserve/wptserve/server.py:883]: Nit: two empty lines between top-level stuff.

File: tools/wptserve/wptserve/server.py
Comment (Hexcles) [inline tools/wptserve/wptserve/server.py:638]: This seems a bit brittle. Are headers in HTTP/2 also case-insensitive? If so, can we do case-insensitive lookup for all headers similar to what you can find in `request.py`? If not, this seems like a bug elsewhere that needs to be fixed.

**Review by Hexcles — COMMENTED** [review]
I reviewed the second commit (WebSockets over HTTP/2 in wptserve) and it looks pretty clean to me (note that I know little about HTTP/2 or WebSockets). Let's wait for the RFC to pass.

File: tools/wptserve/wptserve/ws_h2_handshake.py
Comment (arenevier) [reply tools/wptserve/wptserve/ws_h2_handshake.py:18]: Any thought about that @ricea ?

File: tools/wptserve/wptserve/server.py
Comment (arenevier) [reply tools/wptserve/wptserve/server.py:638]: HTTP/2 headers are case sensitive. I will modify the handshake code to read the correct (lower case) headers instead.

File: tools/wptserve/wptserve/ws_h2_handshake.py
Comment (ricea) [reply tools/wptserve/wptserve/ws_h2_handshake.py:18]: I'd like to do the refactor of pywebsocket after we land this, so we know exactly what we need to export, and can work out the nicest way to export it.

**Comment by arenevier** [comment]
Error on macOS is: "No such file or directory" when trying to launch the browser. So it doesn't seem related to this patch.

https://dev.azure.com/web-platform-tests/wpt/_build/results?buildId=57067&view=logs&j=af75b4de-b26f-5d4d-58e8-b41258076d44&t=052839da-7227-5391-fe2d-4b66ec169a0c

**Comment by Hexcles** [comment]
The RFC has landed, so I'm reviewing the PR now.

There are a few lint/stylistic errors (https://community-tc.services.mozilla.com/tasks/Vw217bFWThSgrwEHQlxqLQ/runs/0/logs/https%3A%2F%2Fcommunity-tc.services.mozilla.com%2Fapi%2Fqueue%2Fv1%2Ftask%2FVw217bFWThSgrwEHQlxqLQ%2Fruns%2F0%2Fartifacts%2Fpublic%252Flogs%252Flive.log#L432). Could you fix them?

File: websockets/constants.js
Comment (Hexcles) [inline websockets/constants.js:10]: Shall we define this at the top of the file for consistency?

**Review by Hexcles — COMMENTED** [review]
Overall the integration pieces LGTM.

One more thing, let's add an HTTP/2 case to `infrastructure/server/wpt-server-websocket.sub.html`. This will prevent the server from regressing.

File: tools/wptserve/wptserve/server.py
Comment (arenevier) [reply tools/wptserve/wptserve/server.py:427]: extract_method_header is in h2 third party library. So on second thought, it seems simpler to run leave it here.

File: tools/wptserve/wptserve/server.py
Comment (arenevier) [reply tools/wptserve/wptserve/server.py]: That sounds good. I have replaced with isomorphic_encode

File: tools/wptserve/wptserve/server.py
Comment (arenevier) [reply tools/wptserve/wptserve/server.py:508]: done

File: websockets/constants.js
Comment (arenevier) [reply websockets/constants.js]: done

**Comment by arenevier** [comment]
> There are a few lint/stylistic errors (https://community-tc.services.mozilla.com/tasks/Vw217bFWThSgrwEHQlxqLQ/runs/0/logs/https%3A%2F%2Fcommunity-tc.services.mozilla.com%2Fapi%2Fqueue%2Fv1%2Ftask%2FVw217bFWThSgrwEHQlxqLQ%2Fruns%2F0%2Fartifacts%2Fpublic%252Flogs%252Flive.log#L432). Could you fix them?

done

> One more thing, let's add an HTTP/2 case to `infrastructure/server/wpt-server-websocket.sub.html`. This will prevent the server from regressing.

I couldn't add a test to wpt-server-websocket.sub.html because it served over http1. I have added a new file http2-websocket.sub.h2.any.js

This test will fail on most current browsers (except I think, Firefox). 

Should I create a ini file in metadata/ ? If so, I'm not quite sure how to fill it.

File: tools/wptserve/wptserve/server.py
Comment (ricea) [inline tools/wptserve/wptserve/server.py:532]: @bencebeky Do you know the answer to this question?

**Review by ricea — APPROVED** [review]
lgtm!

File: tools/wptserve/wptserve/server.py
Comment (bencebeky) [reply tools/wptserve/wptserve/server.py:532]: This is not according to the spec.  According to https://tools.ietf.org/html/rfc8441#section-5, closing the HTTP/2 stream is like closing the TCP connection for regular WebSockets.

I was not aware of this bug, and I did not find any relevant issue on the Chromium bugtracker or any TODO in the source code.  Alas I did not find a test for this behavior either.  I suggest you file a bug at https://crbug.com/new and add a link to that in this comment.

Comment (arenevier) [reply tools/wptserve/wptserve/server.py:532]: I have opened https://bugs.chromium.org/p/chromium/issues/detail?id=1151393

@bencebeky: since this is not according to the spec, and if chromium has a bug, maybe I should remove the TODO and use end_stream here?

Comment (bencebeky) [reply tools/wptserve/wptserve/server.py:532]: Hm actually that makes a lot of sense to me.  Also, thank you for filing the bug.

File: tools/wptserve/wptserve/server.py
Comment (jgraham) [inline tools/wptserve/wptserve/server.py:850]: The docstring setup here is very odd. I think it's better to have the class-level description in the `__init__` docstring rather than having a generic description of what constructing the class does there.

File: tools/wptserve/wptserve/server.py
Comment (jgraham) [inline tools/wptserve/wptserve/server.py:760]: I feel like this should default to None so it has to be explicitly configured.

**Review by jgraham — COMMENTED** [review]
> I couldn't add a test to wpt-server-websocket.sub.html because it served over http1. I have added a new file http2-websocket.sub.h2.any.js
>
> This test will fail on most current browsers (except I think, Firefox).
>
> Should I create a ini file in metadata/ ? If so, I'm not quite sure how to fill it.

In `infrastructure/metadata/infrastructure/server/http2-websocket.sub.h2.any.js.ini`, something like:

```
[http2-websocket.sub.h2.any.html]
  [WSS over h2]
    expected:
      if product != "firefox": FAIL


[http2-websocket.sub.h2.any.worker.html]
  [WSS over h2]
    expected:
      if product != "firefox": FAIL
```

But test this with `wpt run --metadata infrastructure/metadata <browser> infrastructure/server/http2-websocket.sub.h2.any.js`.

File: tools/wptserve/wptserve/ws_h2_handshake.py
Comment (jgraham) [inline tools/wptserve/wptserve/ws_h2_handshake.py:46]: Another docstring to fix here.

**Review by jgraham — APPROVED** [review]
I also haven't looked at the HTTP/2 parts closely, but I assume there's enough expertise that they are already covered.

File: tools/wptserve/wptserve/ws_h2_handshake.py
Comment (arenevier) [reply tools/wptserve/wptserve/ws_h2_handshake.py]: done

**Comment by arenevier** [comment]
I have pushed a new change to this patch:
Since I am adding a **http2** variant, I need the ManifestItem to be considered h2 if the query is equal to **http2**. Without that change, the http2 variants won't be served over h2.

https://github.com/arenevier/wpt/commit/4ffd1c3eea02d9a2c75c6faff32d08e2b3f3735d#diff-8f280b64b0700ab9b4b343adabc7ff4d5ea4b1f45cb6c4bd7f50a19b21ebdb8c

@ricea @jgraham : does that change look good to you?

**Comment by ricea** [comment]
> @ricea @jgraham : does that change look good to you?

Yes, lgtm.

**Comment by jgraham** [comment]
Hmm. I'm not very excited about adding a new kind of filename flag that depends on a query string. That's not something we've had before. I wonder if we should consider a way to make the variant mechanism itself encode additional flags per variant, more like the `any.js` setup.

**Comment by stephenmcgruer** [comment]
@jgraham and I discussed this this morning, and feel that a reasonable compromise is to use a query string, but have it be an actual query parameter e.g. `?flags=h2&...`. There was then further discussion on IRC (https://w3.logbot.info/testing/20201202#c595167):

```
jgraham arno: So I discussed this with smcgruer_[EST] a bit and I think we agree that something like <meta name=variant content="?flags=h2"> would be a good comprimise i.e. we'd reserve the flags query parameter for this purpose
smcgruer_[EST] Could also do `wpt_flags` or something if we wanna namespace it
smcgruer_[EST] Sorry, I have an AI to reply to the RFC with that and have not got to it yet
jgraham smcgruer_[EST] is outsourcing his job to GPT-3
arno jgraham: so it would be flags=h2, and the js would check location.search === '?flags=h2'
arno I will do a patch if you think that's the best solution.
arno In the meantime, what do you think of pushing the current patch (minus the latest addition)? The wptserver part is self contained, so maybe it can be merged already.
jgraham arno: Yeah I think that allows other tests to use this without needing to be exactly like the websockets tests, or hardcoding a bunch of cases
jgraham I think it was approved before that addition? In which case I'm happy for you to push that part :)
arno great. I pushed a new patch.
```

**Comment by arenevier** [comment]
I've added support wpt_flags in https://github.com/arenevier/wpt/commit/977e3ee799aaafc35ca08fc99a0e8dac49935fed

@jgraham @stephenmcgruer @ricea

**Comment by jgraham** [comment]
That looks pretty good to me. I guess there's a question about whether one can pass multiple flags via the query parameter e.g. both `www` and `https`. @gsnedders do you have thoughts on this?

**Comment by arenevier** [comment]
> That looks pretty good to me. I guess there's a question about whether one can pass multiple flags via the query parameter e.g. both `www` and `https`. @gsnedders do you have thoughts on this?

In the current implementation, thanks to how python parse_qs works, you can pass multiple wpt_flags parameters.

`wpt_flags=www&wpt_flags=https`

File: tools/wptserve/wptserve/server.py
Comment (stephenmcgruer) [inline tools/wptserve/wptserve/server.py:502]: It's not clear to me what this `TODO` is for. Is this something that should be fixed in the future (e.g. move to one thread)?

File: tools/manifest/item.py
Comment (stephenmcgruer) [inline tools/manifest/item.py:125]: This is a lint error, from the unittests bot:

`./manifest/item.py:125:24: W503 line break before binary operator`

**Review by stephenmcgruer — APPROVED** [review]
LGTM modulo some nits and small comments. It may have been nice to add some unittests for the `tools/manifest` and `tools/wptserve` changes, given all we have currently is a pair of end-to-end infrastructure tests, but I won't hold up the CL over it.

**Comment by arenevier** [comment]
> LGTM modulo some nits and small comments. It may have been nice to add some unittests for the `tools/manifest` and `tools/wptserve` changes, given all we have currently is a pair of end-to-end infrastructure tests, but I won't hold up the CL over it.

I have added a unittest in tools/manifest. Adding unittests to tools/wptserve seems more tricky.

**Pull request merged**
